src/test/liquidity_tests.cpp

```diff
@@ -0,0 +1,377 @@
+        });
+        BOOST_CHECK(res.ok);
+        BOOST_CHECK(pool.blockCommissionA == 2000000);
+        BOOST_CHECK(pool.reserveA == 298000000);
+        BOOST_CHECK(pool.reserveB == 33534743854);
+    }
+    {
+//        printf("1 COIN (1:1000)\n");
+        CPoolPair pool = *optPool;
+        res = pool.AddLiquidity(COIN, 1000*COIN, {}, [](const CScript &, CAmount liq)-> Res {
+            return Res::Ok();
+        });
+        res = pool.Swap(CTokenAmount{pool.idTokenA, COIN}, PoolPrice{std::numeric_limits<CAmount>::max(), 0}, [&] (CTokenAmount const &ta) -> Res{
+            BOOST_CHECK(ta.nValue == 49773755285);
+            return Res::Ok();
+        });
+        BOOST_CHECK(res.ok);
+        BOOST_CHECK(pool.blockCommissionA == 1000000);
+        BOOST_CHECK(pool.reserveA == 199000000);
+        BOOST_CHECK(pool.reserveB == 50226244715);
+    }
+    {
+//        printf("COIN/1000 (1:1000) (no slope due to commission)\n");
+        CPoolPair pool = *optPool;
+        res = pool.AddLiquidity(COIN, 1000*COIN, {}, [](const CScript &, CAmount liq)-> Res {
+            return Res::Ok();
+        });
+        res = pool.Swap(CTokenAmount{pool.idTokenA, COIN/1000}, PoolPrice{std::numeric_limits<CAmount>::max(), 0}, [&] (CTokenAmount const &ta) -> Res{
+            BOOST_CHECK(ta.nValue == 99000000);
+//              printf("ta = %ld\n", ta.nValue);
+            return Res::Ok();
+        });
+        BOOST_CHECK(res.ok);
+        BOOST_CHECK(pool.blockCommissionA == 1000);
+        BOOST_CHECK(pool.reserveA == 100099000);
+        BOOST_CHECK(pool.reserveB == 99901000000);
+
+//        printf("comissionA = %ld\n", pool.blockCommissionA);
+//        printf("reserveA = %ld\n", pool.reserveA);
+//        printf("reserveB = %ld\n", pool.reserveB);
+    }
+}
+
+void SetPoolRewardPct(CCustomCSView & mnview, DCT_ID idPool, CAmount pct)
+{
+    auto optPool = mnview.GetPoolPair(idPool);
+    BOOST_REQUIRE(optPool);
+    optPool->rewardPct = pct;
+    mnview.SetPoolPair(idPool, *optPool);
+}
+
+void SetPoolTradeFees(CCustomCSView & mnview, DCT_ID idPool, CAmount A, CAmount B)
+{
+    auto optPool = mnview.GetPoolPair(idPool);
+    BOOST_REQUIRE(optPool);
+    optPool->blockCommissionA = A;
+    optPool->blockCommissionB = B;
+    optPool->swapEvent = true;
+    mnview.SetPoolPair(idPool, *optPool);
+}
+
+BOOST_AUTO_TEST_CASE(math_rewards)
+{
+    const int PoolCount = 10; // less than DCT_ID_START!
+    const int ProvidersCount = 10000;
+
+    CCustomCSView mnview(*pcustomcsview);
+
+    // create pools
+    for (int i = 0; i < PoolCount; ++i) {
+        DCT_ID idA, idB, idPool;
+        std::tie(idA, idB, idPool) = CreatePoolNTokens(mnview, "A"+std::to_string(i), "B"+std::to_string(i));
+        auto optPool = mnview.GetPoolPair(idPool);
+        BOOST_REQUIRE(optPool);
+
+    }
+    // create shares
+    mnview.ForEachPoolPair([&] (DCT_ID const & idPool, CPoolPair const & pool) {
+//            printf("pool id = %s\n", idPool.ToString().c_str());
```

Comment:
Remove debug code?

Reply:
Yes. But in a separate PR from this one.